### DISCUSSION TITLE
Syslogs from cisco ASA

### DISCUSSION THREAD

**Discussion by J0hn-W1ick** [Security-Onion-Solutions/securityonion#2945] [Unsupported Versions]
Hello Team! Can you help me! I have 3 weeks trying to parse the logs from my cisco ASA but isn't working
I create a file with a grok pattern that match cisco logs.
I put that file in the custom ingest folder for elasticsearch
Nothing of this had any effect in the syslog dashboard. All the information of the ASA logs appears in the "message" field

**Comment by stevewillia** [comment]
Did you open the ports for the firewall to allow the traffic using so-allow or so-firewall?  You can check by issuing the command:

sudo iptables -L | grep syslog

Did you add an ingest processor for the logs on Elasticsearch?

Check those first.

Steve

**Reply by J0hn-W1ick** [reply]
Thanks Steve.
Yeah, syslog is receiving the logs
I was trying to add the ingest processor without luck, I copied the syslog file from system ingest and modified with a grok pattern to match the ASA logs, then I paste that custom file in the custom ingest folder but is not working. I thing I didn't the custom file correctly, how is the structure of that file?

**Reply by stevewillia** [reply]
Did you add your custom to the global.sls file.  Here is what added for sonicwall.

logstash:
  pipelines:
    search:
      config:
        - custom/9005_sonicwall_output.conf

https://docs.securityonion.net/en/2.3/logstash.html?highlight=Logstash%20

That is where I am stuck.  I cannot get them in to so.

I've developed the parser for Elasticsearch by going to Kibana menu dev tools.  I can show you that if you want.

Steve

**Reply by stevewillia** [reply]
Here is mine.  From looking at the other files it looks like it checks for the ASA logs and ships them to your Elasticsearch parser.

I looked here for examples:

/opt/so/saltstack/default/salt/logstash/pipelines/config/so/

Here is my Logstash shipper to Elasticsearch.  But not working correctly
----------------

filter {
  if [type] == "syslog" or "syslog" in [tags] {
        dissect {
           mapping => {
                mapping => {  "message" =>  "id=%{device} sn=%{serial} %{msg}" }
        }
     }
 }
}

output {
  if [device] =~ "firewall" {
     elasticsearch {
      pipeline => "sonicwall.firewall"
      hosts => "xx.xx.100.252"
      index => "so-sonicwall"
      template_name => "so-firewall"
      template => "/templates/so-firewall-template.json"
      template_overwrite => true
    }
  }
}

fwiw
Steve

**Reply by J0hn-W1ick** [reply]
I didn't add lines to global.sls. should I?
I copy and paste a syslog file from here (/opt/so/saltstack/local/salt/elasticsearch/files/ingest/) to here (/opt/so/conf/elasticsearch/ingest/) and change the grok pattern to:

"<%{POSINT:syslog_pri}>%{DATA:syslog.timestamp}: %%{CISCOTAG:cisco_tag}: %{CISCO_ACTION:action} %{CISCO_XLATE_TYPE:xlate_type} %{WORD:protocol} translation from %{DATA:src_interface}:%{IP:src_ip}/%{INT:src_port}? to %{DATA:dst_interface}:%{IP:dst_ip}/%{INT:dst_port}"

to match ASA logs.
but seems is not doing anything to the logs, my dashboard is still blank. :(

**Reply by stevewillia** [reply]
That's how I read it here:

https://docs.securityonion.net/en/2.3/logstash.html?highlight=Logstash%20

But if you used Kibana dev tools to load the parser, I'm not sure you need to.  But I have done both.

Steve

**Reply by J0hn-W1ick** [reply]
can you show me how you upload the parser  in kibana dev tools?

**Reply by stevewillia** [reply]
Here is mine used for testing.  Just replace my parser and data with yours.

POST _ingest/pipeline/_simulate
    {
  "pipeline": {
  "description": "sonicwall.firewall",
  "processors": [
    {
        "dissect": {
                "field": "message",
                "pattern" : "id=%{device} sn=%{serial} time=\"%{syslog_timestamp}\" fw=%{fw_address} pri=%{pri} c=%{c_id} m=%{msg_id} msg=\"%{rule.name}\" n=%{seq} src=%{source.ip}:%{source.port}:%{fwlan} dst=%{destination.ip}:%{destination.port}:%{fw_eth} proto=%{network.transport}/%{proto} sent=%{bytes} dpi=%{dpi} fw_action=\"%{rule.action}\""
        }
    }
        ]
      },
      "docs": [
        {
          "_source": {
            "message": "id=firewall sn=C0EAE4EAEB50 time=\"2021-02-08 20:43:08 UTC\" fw=216.147.227.82 pri=6 c=262144 m=98 msg=\"Connection Opened\" n=58091987 src=88.88.100.19:58615:X0 dst=168.61.161.212:443:X1 proto=tcp/https sent=60 dpi=0 fw_action=\"NA\""
          }
        }
      ]
    }
  

and the final parser

PUT _ingest/pipeline/sonicwall.firewall
{
  "description": "sonicwall.firewall",
  "processors": [
{
        "dissect": {
                "field": "message",
                "pattern" : "id=%{device} sn=%{serial} time=\"%{syslog_timestamp}\" fw=%{fw_address} pri=%{pri} c=%{c_id} m=%{msg_id} msg=\"%{rule.name}\" n=%{seq} src=%{source.ip}:%{source.port}:%{fwlan} dst=%{destination.ip}:%{destination.port}:%{fw_eth} proto=%{network.transport}/%{proto} sent=%{bytes} dpi=%{dpi} fw_action=\"%{rule.action}\""
        }
},
    {   "rename":{ "field": "id",                "target_field": "event.module",          "ignore_failure": true  } },
    {   "rename":{ "field": "c_id",                "target_field": "host.name",          "ignore_failure": true  } },
        { "set":         { "field": "_index",        "value": "so-firewall", "override": true           } },
        { "set":         { "if": "ctx.network?.transport_id == '0'", "field": "network.transport",        "value": "icmp", "override": true           } },
        {"community_id": { "if": "ctx.network?.transport != null", "field":["source.ip","source.port","destination.ip","destination.port","network.transport"],"target_field":"network.community_id"}},
        { "set":         { "field": "event.module",        "value": "sonicwall", "override": true           } },
     { "set":         { "field": "event.dataset",        "value": "firewall", "override": true           } },
      { "set":         { "field": "host.name",        "value": "firewall", "override": true           } },
      { "set": { "if": "ctx.product != null", "field": "dataset", "value": "{{ product }}", "ignore_failure": true } },
    { "set": { "field": "@timestamp",      "value": "{{ syslog_timestamp }}"       } }

**Reply by stevewillia** [reply]
Here is final parser.  I left out some lines on the previous one.

PUT _ingest/pipeline/sonicwall.firewall
{
  "description": "sonicwall.firewall",
  "processors": [
{
        "dissect": {
                "field": "message",
                "pattern" : "id=%{device} sn=%{serial} time=\"%{syslog_timestamp}\" fw=%{fw_address} pri=%{pri} c=%{c_id} m=%{msg_id} msg=\"%{rule.name}\" n=%{seq} src=%{source.ip}:%{source.port}:%{fwlan} dst=%{destination.ip}:%{destination.port}:%{fw_eth} proto=%{network.transport}/%{proto} sent=%{bytes} dpi=%{dpi} fw_action=\"%{rule.action}\""
        }
},
    {   "rename":{ "field": "id",                "target_field": "event.module",          "ignore_failure": true  } },
    {   "rename":{ "field": "c_id",                "target_field": "host.name",          "ignore_failure": true  } },
        { "set":         { "field": "_index",        "value": "so-firewall", "override": true           } },
        { "set":         { "if": "ctx.network?.transport_id == '0'", "field": "network.transport",        "value": "icmp", "override": true           } },
        {"community_id": { "if": "ctx.network?.transport != null", "field":["source.ip","source.port","destination.ip","destination.port","network.transport"],"target_field":"network.community_id"}},
        { "set":         { "field": "event.module",        "value": "sonicwall", "override": true           } },
     { "set":         { "field": "event.dataset",        "value": "firewall", "override": true           } },
      { "set":         { "field": "host.name",        "value": "firewall", "override": true           } },
      { "set": { "if": "ctx.product != null", "field": "dataset", "value": "{{ product }}", "ignore_failure": true } },
    { "set": { "field": "@timestamp",      "value": "{{ syslog_timestamp }}"       } }



],
  "on_failure" : [
    {
      "set" : {
              "field" : "error.message",
              "value" : "{{ _ingest.on_failure_message }}"
      }
    }
  ]
}

**Reply by stevewillia** [reply]
Or did you mean how to get to dev tools

Open so console -> Kibana -> Menu -> Dev tools

Steve

**Reply by J0hn-W1ick** [reply]
no, I was referring to your first answer. thank you, I'm not use the Dev tools too much, let me try to apply the parser by dev tool and I will tell you the results XD

**Reply by stevewillia** [reply]
Only tricky thing is if your test data has quotes in it.  " or '

You will need to escape the them.  i.e. \"

Otherwise it will give you an error message.

Also everything will stay in dev tools just as you left it even through reboots and shutdowns.

**Reply by stevewillia** [reply]
escape them

\" or \'

**Reply by J0hn-W1ick** [reply]
I add this in the dev tool:

PUT _ingest/pipeline/asa.firewall
{
"description": "ciscoasa.firewall",
"processors": [
{
"dissect": {
"field": "message",
"pattern" : "<%{POSINT:syslog_pri}>%{DATA:syslog.timestamp}: %%{CISCOTAG:cisco_tag}: %{CISCO_ACTION:action} %{CISCO_XLATE_TYPE:xlate_type} %{WORD:protocol} translation from %{DATA:source.interface}:%{IP:source.ip}/%{INT:source.port} to %{DATA:destination.interface}:%{IP:destination.ip}/%{INT:destination.port}"
}
},
{ "set": { "field": "_index", "value": "so-firewall", "override": true } },

{ "set": { "field": "event.module", "value": "cisco asa", "override": true } },
{ "set": { "field": "event.dataset", "value": "firewall", "override": true } },
{ "set": { "field": "host.name", "value": "firewall", "override": true } },
{ "set": { "if": "ctx.product != null", "field": "dataset", "value": "{{ product }}", "ignore_failure": true } },
{ "set": { "field": "@timestamp", "value": "{{ syslog.timestamp }}" } }

],
"on_failure" : [
{
"set" : {
"field" : "error.message",
"value" : "{{ _ingest.on_failure_message }}"
}
}
]
}


but is still not changing anything in logs, should I need restart something?

**Reply by stevewillia** [reply]
Did you write your Logstash shipper that ships the asa logs to your pipeline?  It goes here:

/opt/so/saltstack/local/salt/logstash/pipelines/config/custom/9999_sonicwall

and you need to add the reference the global.sls file

There are examples here
/opt/so/conf/logstash/pipelines/search/

This is where I'm stuck mine aren't getting there either.

I'm looking in to whether the logs actually get to Filebeat.

Firewall looks correct
 iptables -L | grep syslog
ACCEPT     udp  --  anywhere             172.17.0.14          udp dpt:syslog
ACCEPT     udp  --  _gateway             anywhere             udp dpt:syslog
ACCEPT     udp  --  securityonion            anywhere             udp dpt:syslog

I created a tcpdump for docker and looked for syslog traffic at Filebeat and nothing.  I know the firewall is sending the logs.  I checked with ngrep port syslog.

docker run --tty --net=container:so-filebeat tcpdump tcpdump -N -A -v 'udp port 514'
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes

0 packets captured
0 packets received by filter
0 packets dropped by kernel

But if I look at the Wazuh container there is traffic to prove tcpdump is working

docker run --tty --net=container:so-wazuh tcpdump tcpdump -N -A -v 'udp port 1514'
tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
14:36:14.197400 IP (tos 0x0, ttl 64, id 45012, offset 0, flags [DF], proto UDP (17), length 193)
    192.168.100.252.51271 > securityonion-wazuh-manager.1514: UDP, length 165
E.....@.@.....d....	.G.....}#AES:t..._$......R]:........$..d.$.".....o..s...z....*.....*.BP"..SL@?.k..$.o.Mq.c..B.Dr._.[..kt.bg.....&......D.;.P...	.....*...6..-k[nP......Q.....d......E.S,..3R.
14:36:14.198981 IP (tos 0x0, ttl 64, id 45013, offset 0, flags [DF], proto UDP (17), length 257)
    192.168.100.252.51271 > securityonion-wazuh-manager.1514: UDP, length 229
E.....@.@..W..d....	.G......#AES:..	....&/rj.).....J...W...R.~.....c.<.n.7...b.*+Y.Tvs.&0O..dsq..?.F.._-CS..K.Y....<'J....tIm..n....M.K.q.'I.U...[%.I..h....	1Y]c.c.. ,..g.>.X.].....0...RS.p.8.+kn....b...5...]......Zo.;.g..+..n~..".Z.....l...JH...+.y.!.;.u.~
14:36:14.200988 IP (tos 0x0, ttl 64, id 45014, offset 0, flags [DF], proto UDP (17), length 289)
7 packets captured
7 packets received by filter
0 packets dropped by kernel

Steve

**Reply by stevewillia** [reply]
You can check if anything is getting to your pipeline using

so-elasticsearch-pipeline-stats nameofyourpipeline


Look at the counts
Steve

**Reply by J0hn-W1ick** [reply]
let me try that,  do you think is a good option to change to SecOnion 16?

**Reply by stevewillia** [reply]
Yeah if you could try the tcpdump on so-filebeat and see if anything shows up that would be great.

Here is how to setup the docker image

https://rmoff.net/2019/11/29/using-tcpdump-with-docker/

I've been using Elasticsearch, Kibana etc... for years to organize snort, firewall, and windows logs.  I was hoping so 2.3 would make it easier.  Not the case for me.

I've also been using the 16 version for years.  Just setup Filebeat and Logstash to send the logs to Elasticsearch.

Steve

**Reply by J0hn-W1ick** [reply]
sorry man, this was so frustrating, I changed to SecOnion 16 version. did you solve yours?

**Reply by stevewillia** [reply]
Yeah I gave up on trying to get the syslog to docker Filebeat.  I spend a month on it and never got it to work for whatever reason I could not figure out.

I changed the firewall to send it to port 2000 and setup syslog to receive them and store them in a file.  I then setup Filebeat to pick them up and ship them to Logstash and then to Elasticsearch pipeline.

Oh and modified iptables to accept them.

Steve

**Reply by J0hn-W1ick** [reply]
nice, that sounds like a better solution

**Reply by gebhard73** [reply]
Hi, fwiw: we're sending access and syslog logs from ASA to SO 2 by 514/udp w/o having to modify anything under the hood (514 is forwarded to the filebeat docker container, just make sure to do so-allow). Then we modified the syslog parsing and added some fields to the index patterns (see #3175 for some ideas). Perhaps it's worth starting from scratch using a clean SO installation?
Just my 2 cents.
HTH, Gebhard

**Reply by J0hn-W1ick** [reply]
Nice @gebhard73 and are you already parsing ASA logs in SecOnion 2? I mean, it is working?

**Reply by J0hn-W1ick** [reply]
can you share your modified syslog parsing example for ASA?

**Reply by gebhard73** [reply]
I'll ask my employer for permission and will then post it. Pls stand by.

**Reply by J0hn-W1ick** [reply]
sure, no problem, I appreciate your help. anyway if is not possible to share, can you give me a clue about how to build the file? because right now I have no clue XD

**Reply by gebhard73** [reply]
meanwhile I've the OK from my employer and will post the syslog config file ... I've to do a little fine tuning and cleaning and will come back then

**Reply by J0hn-W1ick** [reply]
awesome! I think I and more people here will appreciate that, will be very helpful

**Reply by gebhard73** [reply]
As promised pls find my suggestion here: #3547

**Comment by sinnetll19** [comment]
can you please share the Grok Pattern that you created for the ASA. i am also trying to get those logs into SO.

**Reply by gebhard73** [reply]
Yes, as said above, I'll share it as soon as it gets stable and reliable... It'll be for the access logs for ASA.

**Reply by gebhard73** [reply]
As promised pls find my suggestion here: #3547
